test/test_query_geo.cpp

```diff
@@ -180,4 +187,66 @@ TEST(Geospatial_GeoWithinShapes)
     }
 }
 
+TEST(Geospatial_PolygonValidation)
+{
+    Group g;
+    std::vector<Geospatial> points = {GeoPoint{40.7128, -74.0060}, GeoPoint{55.6761, 12.5683},
+                                      GeoPoint{55.6280, 12.0826}};
+    TableRef table = setup_with_points(g, points);
+    ColKey location_column_key = table->get_column_key("location");
+    Geospatial geo_poly{GeoPolygon{{GeoPoint{40.7128, -74.006}, GeoPoint{55.6761, 12.5683}, GeoPoint{55.628, 12.0826},
+                                    GeoPoint{40.7128, -74.006}}}};
+    CHECK(geo_poly.is_valid().is_ok());
+    Query query = table->column<Link>(location_column_key).geo_within(geo_poly);
+    CHECK_EQUAL(query.count(), 1);
+
+    Geospatial poly_mismatch_loop{GeoPolygon{{GeoPoint{40.7128, -74.006}, GeoPoint{55.6761, 12.5683},
+                                              GeoPoint{55.628, 12.0826}, GeoPoint{40.7128, -74.000}}}};
+    Status status = poly_mismatch_loop.is_valid();
+    CHECK(!status.is_ok());
+    CHECK_EQUAL(status.reason(), "Loop is not closed, first vertex 'GeoPoint([40.7128, -74.006])' does not equal "
+                                 "last vertex 'GeoPoint([40.7128, -74])'");
+
+    Geospatial poly_three_point{
+        GeoPolygon{{GeoPoint{40.7128, -74.006}, GeoPoint{55.6761, 12.5683}, GeoPoint{40.7128, -74.006}}}};
+    status = poly_three_point.is_valid();
+    CHECK(!status.is_ok());
+    CHECK_EQUAL(status.reason(), "Loop 0 must have at least 3 different vertices, 2 unique vertices were provided");
+
+    Geospatial loop_outside{
+        GeoPolygon{{{GeoPoint{40.7128, -74.006}, GeoPoint{55.6761, 12.5683}, GeoPoint{55.628, 12.0826},
+                     GeoPoint{40.7128, -74.006}},
+                    {GeoPoint{39, -74.006}, GeoPoint{56, 12.5683}, GeoPoint{56, 12.0826}, GeoPoint{39, -74.006}}}}};
+    status = loop_outside.is_valid();
+    CHECK(!status.is_ok());
+    CHECK_EQUAL(
+        status.reason(),
+        "Secondary loop 1 not contained by first exterior loop - secondary loops must be holes in the first loop");
```

Comment:
Looks like S2 uses "loops" but the GeoJson spec talks about "rings" and so do the MongoDB docs. If we align the SDKs to that terminology, it'd be a good idea for the core error messages to be aligned as well.

Reply:
Good call, I've updated the error messages to use "ring".